File: Classes/@Component/Component.m
classdef Component
% Component  Pure-component thermodynamic properties.
%
%   Typical usage: load from database
%     comp = Component.fromDatabase('CH4');
%     comps = Component.fromDatabaseArray({'CH4', 'C2H6'});
%
%   Or construct directly:
%     comp = Component(name, formula, MW, Tc, Pc, Vc, Zc,
%                      acentric_factor, Psat_eq, Psat_coefs,
%                      PsatTrange, Psatrange,
%                      dh_vap_eq, dh_vap_coefs, dh_vap_Trange, dh_vap_range,
%                      cp_liq_eq, cp_liq_coefs, cp_liq_Trange, cp_liq_range,
%                      cp_ig_eq,  cp_ig_coefs,  cp_ig_Trange,  cp_ig_range,
%                      dhf_ig, dgf_ig, ds_ig, dh_comb)

%{
Copyright (c) 2012, 2013, Ali Akbar Eftekhari
All rights reserved.

Redistribution and use in source and binary forms, with or
without modification, are permitted provided that the following
conditions are met:

    *   Redistributions of source code must retain the above copyright notice,
        this list of conditions and the following disclaimer.
    *   Redistributions in binary form must reproduce the above
        copyright notice, this list of conditions and the following
        disclaimer in the documentation and/or other materials provided
        with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
%}

    properties
        name
        formula
        MW
        Tc
        Pc
        Vc
        Zc
        acentric_factor
        Psat_eq
        Psat_coefs
        PsatTrange
        Psatrange
        dh_vap_eq
        dh_vap_coefs
        dh_vap_Trange
        dh_vap_range
        cp_liq_eq
        cp_liq_coefs
        cp_liq_Trange
        cp_liq_range
        cp_ig_eq
        cp_ig_coefs
        cp_ig_Trange
        cp_ig_range
        dhf_ig
        dgf_ig
        ds_ig
        dh_comb
    end

    methods
        function obj = Component(name, formula, ...
                MW, Tc, Pc, Vc, Zc, acentric_factor, ...
                Psat_eq, Psat_coefs, PsatTrange, Psatrange, ...
                dh_vap_eq, dh_vap_coefs, dh_vap_Trange, dh_vap_range, ...
                cp_liq_eq, cp_liq_coefs, cp_liq_Trange, cp_liq_range, ...
                cp_ig_eq, cp_ig_coefs, cp_ig_Trange, cp_ig_range, ...
                dhf_ig, dgf_ig, ds_ig, dh_comb)
            % Zero-argument call is used by Octave/MATLAB to pre-allocate
            % object arrays (e.g. obj(n) = Component(...)).
            if nargin == 0; return; end
            obj.name            = name;
            obj.formula         = formula;
            obj.MW              = MW;
            obj.Tc              = Tc;
            obj.Pc              = Pc;
            obj.Vc              = Vc;
            obj.Zc              = Zc;
            obj.acentric_factor = acentric_factor;
            obj.Psat_eq         = Psat_eq;
            obj.Psat_coefs      = Psat_coefs;
            obj.PsatTrange      = PsatTrange;
            obj.Psatrange       = Psatrange;
            obj.dh_vap_eq       = dh_vap_eq;
            obj.dh_vap_coefs    = dh_vap_coefs;
            obj.dh_vap_Trange   = dh_vap_Trange;
            obj.dh_vap_range    = dh_vap_range;
            obj.cp_liq_eq       = cp_liq_eq;
            obj.cp_liq_coefs    = cp_liq_coefs;
            obj.cp_liq_Trange   = cp_liq_Trange;
            obj.cp_liq_range    = cp_liq_range;
            obj.cp_ig_eq        = cp_ig_eq;
            obj.cp_ig_coefs     = cp_ig_coefs;
            obj.cp_ig_Trange    = cp_ig_Trange;
            obj.cp_ig_range     = cp_ig_range;
            obj.dhf_ig          = dhf_ig;
            obj.dgf_ig          = dgf_ig;
            obj.ds_ig           = ds_ig;
            obj.dh_comb         = dh_comb;
        end

        function p_sat = vapor_pressure(obj, T)
            % vapor_pressure(T)  returns p_sat [Pa] at temperature T [K]
            p_sat = obj.Psat_eq(T, obj.Psat_coefs);
        end

        function dh_v = dh_vap(obj, T)
            % dh_vap(T)  returns enthalpy of vaporisation [J/mol] at T [K]
            dh_v = obj.dh_vap_eq(T/obj.Tc, obj.dh_vap_coefs);
        end

        function cp_l = cp_liq(obj, T)
            % cp_liq(T)  returns liquid heat capacity [J/(mol·K)] at T [K]
            cp_l = obj.cp_liq_eq(T/obj.Tc, obj.cp_liq_coefs);
        end

        function cp = cp_ig(obj, T)
            % cp_ig(T)  returns ideal-gas heat capacity [J/(mol·K)] at T [K]
            cp = obj.cp_ig_eq(T, obj.cp_ig_coefs);
        end
    end

    methods (Static)
        function comp = fromDatabase(name_or_formula)
            % fromDatabase  Load a single Component from puredata.mat.
            %   comp = Component.fromDatabase('CH4')
            %   comp = Component.fromDatabase('methane')
            comps = Component.fromDatabaseArray({name_or_formula});
            if isempty(comps)
                error('Component:notFound', ...
                    'Component "%s" not found in database.', name_or_formula);
            end
            comp = comps(1);
        end

        function [components, flag] = fromDatabaseArray(name_formula_cell)
            % fromDatabaseArray  Load multiple Components from puredata.mat.
            %   [comps, flag] = Component.fromDatabaseArray({'CH4','C2H6'})
            %   flag contains indices of names that were not found.
            load puredata.mat
            N = length(name_formula_cell);
            data_index = zeros(1, N);
            for n = 1:N
                idx = find(strcmpi(component_formula, name_formula_cell(n)), 1);
                if ~isempty(idx)
                    data_index(n) = idx;
                else
                    idx = find(strcmpi(component_name, name_formula_cell(n)), 1);
                    if ~isempty(idx)
                        data_index(n) = idx;
                    end
                end
            end
            flag = find(data_index == 0);
            data_index = data_index(data_index > 0);
            Nfound = length(data_index);
            if Nfound == 0
                components = [];
                return
            end
            % Pre-allocate by constructing the last element first so MATLAB
            % knows the array type, then fill in order.
            i = Nfound;
            ci = data_index(i);
            if liquid_heat_capacity_equation_num(ci) == 1
                cp_liq_fn = liquid_heat_capacity_equation_1;
            else
                cp_liq_fn = liquid_heat_capacity_equation_2;
            end
            if ideal_gas_heat_capacity_equation_num(ci) == 1
                cp_ig_fn = ideal_gas_heat_capacity_equation_1;
            elseif ideal_gas_heat_capacity_equation_num(ci) == 2
                cp_ig_fn = ideal_gas_heat_capacity_equation_2;
            else
                cp_ig_fn = ideal_gas_heat_capacity_equation_3;
            end
            components(Nfound) = Component( ...
                component_name{ci}, component_formula{ci}, ...
                Molecular_weight_data(ci), ...
                critical_temperature_data(ci), critical_pressure_data(ci), ...
                critical_volume_data(ci), critical_compressibility_factor(ci), ...
                acentric_factor(ci), ...
                vapor_pressure_equation, vapor_pressure_coefs(ci,:), ...
                vapor_pressure_T_range(ci,:), vapor_pressure_p_range(ci,:), ...
                dh_vaporization_equation, dh_vaporization_coefs(ci,:), ...
                dh_vaporization_T_range(ci,:), dh_vaporization_dhv_range(ci,:), ...
                cp_liq_fn, liquid_heat_capacity_coef(ci,:), ...
                liquid_heat_capacity_T_range(ci,:), liquid_heat_capacity_cp_range(ci,:), ...
                cp_ig_fn, ideal_gas_heat_capacity_coefs(ci,:), ...
                ideal_gas_heat_capacity_T_range(ci,:), ideal_gas_heat_capacity_cp_range(ci,:), ...
                dh_formation_ig(ci), dg_formation_ig(ci), ...
                ds_ideal_gas(ci), dh_combustion(ci));
            for n = 1:Nfound-1
                ci = data_index(n);
                if liquid_heat_capacity_equation_num(ci) == 1
                    cp_liq_fn = liquid_heat_capacity_equation_1;
                else
                    cp_liq_fn = liquid_heat_capacity_equation_2;
                end
                if ideal_gas_heat_capacity_equation_num(ci) == 1
                    cp_ig_fn = ideal_gas_heat_capacity_equation_1;
                elseif ideal_gas_heat_capacity_equation_num(ci) == 2
                    cp_ig_fn = ideal_gas_heat_capacity_equation_2;
                else
                    cp_ig_fn = ideal_gas_heat_capacity_equation_3;
                end
                components(n) = Component( ...
                    component_name{ci}, component_formula{ci}, ...
                    Molecular_weight_data(ci), ...
                    critical_temperature_data(ci), critical_pressure_data(ci), ...
                    critical_volume_data(ci), critical_compressibility_factor(ci), ...
                    acentric_factor(ci), ...
                    vapor_pressure_equation, vapor_pressure_coefs(ci,:), ...
                    vapor_pressure_T_range(ci,:), vapor_pressure_p_range(ci,:), ...
                    dh_vaporization_equation, dh_vaporization_coefs(ci,:), ...
                    dh_vaporization_T_range(ci,:), dh_vaporization_dhv_range(ci,:), ...
                    cp_liq_fn, liquid_heat_capacity_coef(ci,:), ...
                    liquid_heat_capacity_T_range(ci,:), liquid_heat_capacity_cp_range(ci,:), ...
                    cp_ig_fn, ideal_gas_heat_capacity_coefs(ci,:), ...
                    ideal_gas_heat_capacity_T_range(ci,:), ideal_gas_heat_capacity_cp_range(ci,:), ...
                    dh_formation_ig(ci), dg_formation_ig(ci), ...
                    ds_ideal_gas(ci), dh_combustion(ci));
            end
        end
    end
end

File: pvt/component.py
"""Component — pure-component thermodynamic properties.

Each Component holds critical properties, correlations for vapor pressure,
enthalpy of vaporization, liquid and ideal-gas heat capacity, and formation
data. Components are immutable value objects.
"""

from dataclasses import dataclass, field

import numpy as np

from ._data import get_component_data, get_components_data
from ._equations import EQUATION_FUNCTIONS


@dataclass(frozen=True)
class Component:
    """Pure-component thermodynamic properties container.

    All properties are in SI units: temperature [K], pressure [Pa], energy [J],
    heat capacity [J/(mol·K)], volume [m³/mol].

    Attributes:
        name: Component name (e.g. 'Methane').
        formula: Chemical formula (e.g. 'CH4').
        MW: Molecular weight [kg/mol].
        Tc: Critical temperature [K].
        Pc: Critical pressure [Pa].
        Vc: Critical volume [m³/mol].
        Zc: Critical compressibility factor.
        acentric_factor: Pitzer acentric factor.
        Psat_eq: Vapor pressure equation type identifier.
        Psat_coefs: Vapor pressure coefficients (1D array).
        Psat_Trange: Valid temperature range [Tmin, Tmax] for Psat.
        Psat_range: Pressure range corresponding to Psat_Trange.
        dh_vap_eq: Enthalpy of vaporization equation type.
        dh_vap_coefs: dh_vap coefficients (1D array).
        dh_vap_Trange: Valid T range [Tmin, Tmax] for dh_vap.
        dh_vap_range: dh_vap values at Trange bounds.
        cp_liq_eq: Liquid heat capacity equation type.
        cp_liq_coefs: cp_liq coefficients (1D array).
        cp_liq_Trange: Valid T range for cp_liq.
        cp_liq_range: cp_liq values at Trange bounds.
        cp_ig_eq: Ideal-gas heat capacity equation type.
        cp_ig_coefs: cp_ig coefficients (1D array).
        cp_ig_Trange: Valid T range for cp_ig.
        cp_ig_range: cp_ig values at Trange bounds.
        dhf_ig: Ideal-gas enthalpy of formation [J/mol].
        dgf_ig: Ideal-gas Gibbs free energy of formation [J/mol].
        ds_ig: Ideal-gas absolute entropy [J/(mol·K)].
        dh_comb: Enthalpy of combustion [J/mol].
    """

    name: str
    formula: str
    MW: float
    Tc: float
    Pc: float
    Vc: float
    Zc: float
    acentric_factor: float
    Psat_eq: str
    Psat_coefs: np.ndarray = field(default_factory=lambda: np.zeros(5))
    Psat_Trange: np.ndarray = field(default_factory=lambda: np.zeros(2))
    Psat_range: np.ndarray = field(default_factory=lambda: np.zeros(2))
    dh_vap_eq: str = ""
    dh_vap_coefs: np.ndarray = field(default_factory=lambda: np.zeros(4))
    dh_vap_Trange: np.ndarray = field(default_factory=lambda: np.zeros(2))
    dh_vap_range: np.ndarray = field(default_factory=lambda: np.zeros(2))
    cp_liq_eq: str = ""
    cp_liq_coefs: np.ndarray = field(default_factory=lambda: np.zeros(5))
    cp_liq_Trange: np.ndarray = field(default_factory=lambda: np.zeros(2))
    cp_liq_range: np.ndarray = field(default_factory=lambda: np.zeros(2))
    cp_ig_eq: str = ""
    cp_ig_coefs: np.ndarray = field(default_factory=lambda: np.zeros(5))
    cp_ig_Trange: np.ndarray = field(default_factory=lambda: np.zeros(2))
    cp_ig_range: np.ndarray = field(default_factory=lambda: np.zeros(2))
    dhf_ig: float = 0.0
    dgf_ig: float = 0.0
    ds_ig: float = 0.0
    dh_comb: float = 0.0

    def vapor_pressure(self, T: float) -> float:
        """Compute vapor pressure at temperature T [K].

        Returns p_sat in [Pa].
        """
        fn = EQUATION_FUNCTIONS[self.Psat_eq]
        return fn(T, self.Psat_coefs)

    def dh_vap(self, T: float) -> float:
        """Compute enthalpy of vaporization at temperature T [K].

        Returns dh_vap in [J/mol].
        """
        fn = EQUATION_FUNCTIONS[self.dh_vap_eq]
        return fn(T / self.Tc, self.dh_vap_coefs)

    def cp_liq(self, T: float) -> float:
        """Compute liquid heat capacity at temperature T [K].

        Returns cp in [J/(mol·K)].
        """
        fn = EQUATION_FUNCTIONS[self.cp_liq_eq]
        if "dippr100" in self.cp_liq_eq:
            return fn(T / self.Tc, self.cp_liq_coefs)
        return fn(T, self.cp_liq_coefs)

    def cp_ig(self, T: float) -> float:
        """Compute ideal-gas heat capacity at temperature T [K].

        Returns cp in [J/(mol·K)].
        """
        fn = EQUATION_FUNCTIONS[self.cp_ig_eq]
        return fn(T, self.cp_ig_coefs)

    @staticmethod
    def from_database(name_or_formula: str) -> "Component":
        """Load a single component from the built-in database.

        Args:
            name_or_formula: Component name (e.g. 'Methane') or formula (e.g. 'CH4').

        Returns:
            Component instance.

        Raises:
            KeyError: If the component is not found in the database.
        """
        data = get_component_data(name_or_formula)
        if data is None:
            raise KeyError(f"Component '{name_or_formula}' not found in database.")
        return Component._from_dict(data)

    @staticmethod
    def from_database_array(names_or_formulas: list[str]) -> tuple[list["Component"], list[int]]:
        """Load multiple components from the built-in database.

        Args:
            names_or_formulas: List of component names or formulas.

        Returns:
            Tuple of (components, not_found_indices).
        """
        data_list, not_found = get_components_data(names_or_formulas)
        components = [Component._from_dict(d) for d in data_list]
        return components, not_found

    @staticmethod
    def _from_dict(data: dict) -> "Component":
        """Create a Component from a database dictionary."""
        return Component(
            name=data["name"],
            formula=data["formula"],
            MW=float(data["MW"]),
            Tc=float(data["Tc"]),
            Pc=float(data["Pc"]),
            Vc=float(data["Vc"]),
            Zc=float(data["Zc"]),
            acentric_factor=float(data["acentric_factor"]),
            Psat_eq=data["Psat_eq"],
            Psat_coefs=np.array(data["Psat_coefs"]),
            Psat_Trange=np.array(data["Psat_Trange"]),
            Psat_range=np.array(data["Psat_range"]),
            dh_vap_eq=data["dh_vap_eq"],
            dh_vap_coefs=np.array(data["dh_vap_coefs"]),
            dh_vap_Trange=np.array(data["dh_vap_Trange"]),
            dh_vap_range=np.array(data["dh_vap_range"]),
            cp_liq_eq=data["cp_liq_eq"],
            cp_liq_coefs=np.array(data["cp_liq_coefs"]),
            cp_liq_Trange=np.array(data["cp_liq_Trange"]),
            cp_liq_range=np.array(data["cp_liq_range"]),
            cp_ig_eq=data["cp_ig_eq"],
            cp_ig_coefs=np.array(data["cp_ig_coefs"]),
            cp_ig_Trange=np.array(data["cp_ig_Trange"]),
            cp_ig_range=np.array(data["cp_ig_range"]),
            dhf_ig=float(data["dhf_ig"]),
            dgf_ig=float(data["dgf_ig"]),
            ds_ig=float(data["ds_ig"]),
            dh_comb=float(data["dh_comb"]),
        )
